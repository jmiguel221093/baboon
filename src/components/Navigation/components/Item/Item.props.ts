import type { MouseEvent } from "react";
import type { IconType } from "../../../../types";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type SubNavigationItemType<TAnchorType = any> = {
	/** Indicates weather the link is disabled */
	disabled?: boolean;
	/** The text for the link */
	label?: string;
	/** Indicates weather the link is selected */
	selected?: boolean;
	/** Callback triggered when the item has been pressed */
	onClick?: (event?: MouseEvent) => void;
	/** Sets the link text truncated */
	truncateText?: boolean;
	/** The props that will be passed to the anchor component */
	anchorProps?: TAnchorType;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type NavigationSectionItemProps<TAnchorType = any> = {
	/** The icon of the link */
	icon?: React.ReactNode | IconType;
	/** The text on the badge */
	badge?: string | null;
	/** The text for the link */
	label?: string;
	/** Indicates weather the link is disabled */
	disabled?: boolean;
	/** Indicates weather the link is selected */
	selected?: boolean;
	/** Callback triggered when the item has been pressed */
	onClick?: (event?: MouseEvent) => void;
	/** Sets the link text truncated */
	truncateText?: boolean;
	/** The props that will be passed to the anchor component */
	anchorProps?: TAnchorType;
	/** The sub items of the link */
	subItems?: SubNavigationItemType[];
};
