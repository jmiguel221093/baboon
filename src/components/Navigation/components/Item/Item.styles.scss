@import "../../../../styles/main.scss";

$navigation-section-item-padding: var(--spacing-4);
$navigation-section-item-border-radius: var(--border-radius-9);
$navigation-section-item-hover-background: var(
	--color-background-surface-disabled
);
$navigation-section-item-active-background: var(
	--color-background-surface-neutral-focus
);
$navigation-section-item-selected-background: var(
	--color-background-surface-primary
);

$navigation-section-item-font-size: var(--font-size-75);
$navigation-section-item-line-height: var(--font-line-height-3);
$navigation-section-item-selected-font-color: var(--color-text-on-dark);
$navigation-section-item-selected-icon-color: var(--color-icon-on-dark);
$navigation-section-item-disabled-font-color: var(--color-text-disabled);
$navigation-section-item-disabled-icon-color: var(--color-base-gray-400);

$navigation-section-item-content-margin: var(--spacing-5);

$navigation-section-item-motion-duration: var(--motion-duration-250);
$navigation-section-item-motion-ease: var(--motion-ease-in-out);

$navigation-section-item-subitem-unselected-color: var(--color-text-subtle);
$navigation-section-item-subitem-font-weight: var(--font-weight-medium);
$navigation-section-item-subitem-border-radius: var(--border-radius-5);
$navigation-section-item-subitem-vertical-padding: calc(
	var(--spacing-1) + var(--spacing-05)
);
$navigation-section-item-subitem-horizontal-padding: calc(
	var(--spacing-12) + var(--spacing-05)
);
$navigation-section-item-subitem-selected-border-color: var(
	--color-background-surface-primary
);
$navigation-section-item-subitem-border-spacing: var(--spacing-1);
$navigation-section-item-subitem-border-border-radius: var(
	--border-radius-rounded
);

.Navigation__Item {
	display: block;
	width: 100%;
	a {
		display: inline-flex;
		align-items: center;
		width: 100%;
		text-decoration: none;
		font-size: $navigation-section-item-font-size;
		line-height: $navigation-section-item-line-height;
		font-weight: 600;
		padding: $navigation-section-item-padding;
		color: inherit;
		border-radius: $navigation-section-item-border-radius;
		transition: background-color $navigation-section-item-motion-duration
			$navigation-section-item-motion-ease;
		&:hover {
			background-color: $navigation-section-item-hover-background;
		}
		&:active {
			background-color: $navigation-section-item-active-background;
		}
	}
	&__IconWrapper {
		line-height: 0;
		&:not(:last-child) {
			margin-right: $navigation-section-item-content-margin;
		}
	}
	&__Text {
		flex: 1;
		&:not(:last-child) {
			margin-right: $navigation-section-item-content-margin;
		}
	}
	&--selected {
		& > a {
			background-color: $navigation-section-item-selected-background !important;
		}
		.Navigation__Item {
			&__Text {
				color: $navigation-section-item-selected-font-color !important;
			}
			&__IconWrapper svg {
				color: $navigation-section-item-selected-icon-color !important;
			}
		}
	}
	&--disabled {
		& > a {
			background-color: transparent !important;
			color: $navigation-section-item-disabled-font-color !important;
			pointer-events: none;
		}
		.Navigation__Item {
			&__IconWrapper svg {
				color: $navigation-section-item-disabled-icon-color !important;
			}
		}
	}
	&#{&}--collapsed {
		.Navigation__Item {
			&__Text,
			&__BadgeWrapper {
				display: none;
			}
		}
	}
	&#{&}--truncateText {
		.Navigation__Item {
			&__Text {
				white-space: nowrap;
				overflow: hidden;
				text-overflow: ellipsis;
			}
		}
	}
	&__SubItem {
		position: relative;
		display: flex;
		align-items: stretch;
		margin-top: $navigation-section-item-subitem-border-spacing;
		width: 100%;
		&:before {
			flex: none;
			display: block;
			width: 2px;
			content: "";
			background-color: transparent;
			margin-right: $navigation-section-item-subitem-border-spacing;
			border-radius: $navigation-section-item-subitem-border-border-radius;
		}
		& > a {
			flex: 1;
			padding: $navigation-section-item-subitem-vertical-padding
				$navigation-section-item-padding
				$navigation-section-item-subitem-vertical-padding
				$navigation-section-item-subitem-horizontal-padding;
			border-radius: $navigation-section-item-subitem-border-radius;
			font-weight: $navigation-section-item-subitem-font-weight;
			color: $navigation-section-item-subitem-unselected-color;
			background: transparent !important;
			width: 0;
			&:hover {
				background-color: $navigation-section-item-hover-background !important;
			}
		}
		&#{&}--truncateText {
			.Navigation__Item__SubItem {
				&__Text {
					white-space: nowrap;
					overflow: hidden;
					text-overflow: ellipsis;
				}
			}
		}
		&#{&}--selected {
			& > a {
				background-color: $navigation-section-item-hover-background !important;
				color: inherit;
			}
			&:before {
				background-color: $navigation-section-item-selected-background;
			}
		}
		&#{&}--disabled {
			& > a {
				pointer-events: none;
				color: $navigation-section-item-disabled-font-color !important;
			}
		}
	}
}
