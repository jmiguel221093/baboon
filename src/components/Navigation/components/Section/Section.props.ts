import type { BasicIconActionType } from "../../../../types";
import type { NavigationSectionItemProps } from "../Item/Item.props";

export interface NavigationSectionProps {
	/** Items for the links list */
	items?: NavigationSectionItemProps[];
	/** The title of the section */
	title?: string;
	/** The action used in the section divider*/
	action?: BasicIconActionType;
}
