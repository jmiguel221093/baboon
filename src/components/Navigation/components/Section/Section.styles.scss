$navigation-section-inner-spacing: var(--spacing-1);
$navigation-section-border-color: var(--color-border-disabled);

$navigation-section-border-width: calc(100% - var(--spacing-4) * 2);
$navigation-section-spacing: var(--spacing-8);

$navigation-section-title-padding: var(--spacing-4);
$navigation-section-title-margin-bottom: var(--spacing-4);

.Navigation__Section {
	&:not(:first-child) {
		margin-top: $navigation-section-spacing;
		&:before {
			content: "";
			display: block;
			height: 0;
			width: $navigation-section-border-width;
			margin: auto;
			border-top: solid 1px $navigation-section-border-color;
			margin-bottom: $navigation-section-spacing;
		}
	}
	&__DividerWrapper {
		text-transform: uppercase;
		padding-left: $navigation-section-title-padding;
		padding-right: $navigation-section-title-padding;
		margin-bottom: $navigation-section-title-margin-bottom;
		display: flex;
		align-items: center;
	}
	&__TitleWrapper {
		flex: 1;
	}
	ul {
		& > * {
			&:not(:last-child) {
				margin-bottom: $navigation-section-inner-spacing;
			}
		}
	}
	&#{&}--collapsed {
		.Navigation__Section {
			&__DividerWrapper,
			&__TitleWrapper {
				display: none;
			}
		}
	}
}
