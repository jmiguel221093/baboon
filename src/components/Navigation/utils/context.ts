import { createContext } from "react";

export interface NavigationContextType {
	anchorComponent?: React.FunctionComponent;
	collapsed?: boolean;
}

const NavigationContext = createContext<NavigationContextType>({});

export default NavigationContext;
