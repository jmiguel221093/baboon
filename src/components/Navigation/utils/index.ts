import NavigationContext from "./context";

export * from "./context";
export { NavigationContext };
