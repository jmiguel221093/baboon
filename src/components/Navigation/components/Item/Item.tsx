import { useContext } from "react";
import { cssClassNames, isValidIcon } from "../../../../utils";
import { Icon } from "../../../Icon";
import type {
	NavigationSectionItemProps,
	SubNavigationItemType,
} from "./Item.props";

import "./Item.styles.scss";
import { NavigationContext } from "../../utils";
import type { NavigationContextType } from "../../utils";
import { Badge } from "../../../Badge";

const COMPONENT_NAME = "Navigation__Item";

const Item = ({
	anchorProps,
	badge,
	disabled,
	icon,
	label,
	onClick,
	selected,
	truncateText,
	subItems = [],
}: NavigationSectionItemProps) => {
	const { anchorComponent, collapsed } =
		useContext<NavigationContextType>(NavigationContext);

	const className = cssClassNames(
		COMPONENT_NAME,
		selected && "selected",
		disabled && "disabled",
		collapsed && "collapsed",
		truncateText && "truncateText"
	);

	const AnchorComponent = anchorComponent || "a";

	const renderSubItem = (
		subItem: SubNavigationItemType,
		index: number
	): React.ReactNode => {
		const {
			anchorProps: subItemAnchorProps,
			disabled: subItemDisabled,
			label: subItemLabel,
			onClick: subItemOnClick,
			selected: subItemSelected,
			truncateText: subItemTruncateText,
		} = subItem;
		const subItemClassName = cssClassNames(
			`${COMPONENT_NAME}__SubItem`,
			subItemSelected && "selected",
			subItemDisabled && "disabled",
			subItemTruncateText && "truncateText"
		);
		return (
			<li key={index} className={subItemClassName}>
				<AnchorComponent
					{...subItemAnchorProps}
					onClick={subItemOnClick}
				>
					<span className={`${COMPONENT_NAME}__SubItem__Text`}>
						{subItemLabel}
					</span>
				</AnchorComponent>
			</li>
		);
	};

	const subItemsMarkup =
		subItems.length && selected && !collapsed ? (
			<ul className={`${COMPONENT_NAME}__SubItems`}>
				{subItems.map(renderSubItem)}
			</ul>
		) : null;

	const iconSource = isValidIcon(icon) ? (
		<Icon source={icon} size="tiny" color="neutral" />
	) : (
		<>{icon}</>
	);

	const iconMarkup = icon ? (
		<div className={`${COMPONENT_NAME}__IconWrapper`}>
			{iconSource as React.ReactNode}
		</div>
	) : null;

	const labelMarkup = label ? (
		<span className={`${COMPONENT_NAME}__Text`}>{label}</span>
	) : null;

	const badgeMarkup = badge ? (
		<div className={`${COMPONENT_NAME}__BadgeWrapper`}>
			<Badge>{badge}</Badge>
		</div>
	) : null;

	return (
		<li className={className}>
			<AnchorComponent {...anchorProps} onClick={onClick}>
				{iconMarkup}
				{labelMarkup}
				{badgeMarkup}
			</AnchorComponent>
			{subItemsMarkup}
		</li>
	);
};

export default Item;
