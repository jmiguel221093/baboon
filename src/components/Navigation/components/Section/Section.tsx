import { useContext } from "react";
import { cssClassNames } from "../../../../utils";
import { Button } from "../../../Button";
import { Text } from "../../../Text";
import { NavigationContext } from "../../utils";
import type { NavigationContextType } from "../../utils";

import { Item } from "../Item";
import type { NavigationSectionItemProps } from "../Item";
import type { NavigationSectionProps } from "./Section.props";

import "./Section.styles.scss";

const COMPONENT_NAME = "Navigation__Section";

const Section = ({ items, title, action }: NavigationSectionProps) => {
	const { collapsed } = useContext<NavigationContextType>(NavigationContext);

	const className = cssClassNames(COMPONENT_NAME, collapsed && "collapsed");

	const renderNavigationItem = (
		item: NavigationSectionItemProps,
		index: number
	): React.ReactNode => <Item key={index} {...item} />;

	const itemsMarkup = items ? (
		<ul className={`${COMPONENT_NAME}__Items`}>
			{items.map(renderNavigationItem)}
		</ul>
	) : null;

	const dividerMarkup =
		title || action ? (
			<div className={`${COMPONENT_NAME}__DividerWrapper`}>
				<div className={`${COMPONENT_NAME}__TitleWrapper`}>
					{title && (
						<Text
							variant="caption"
							color="neutral"
							fontWeight="semibold"
						>
							{title}
						</Text>
					)}
				</div>
				{action && (
					<div className={`${COMPONENT_NAME}__ActionWrapper`}>
						<Button
							id={action.id}
							icon={action.icon}
							clean
							onClick={action.onPress}
							disabled={action.disabled}
							onMouseEnter={action.onMouseEnter}
							onTouchStart={action.onTouchStart}
							loading={action.loading}
							danger={action.danger}
							size="extra-small"
						/>
					</div>
				)}
			</div>
		) : null;

	return (
		<div className={className}>
			{dividerMarkup}
			{itemsMarkup}
		</div>
	);
};

export default Section;
