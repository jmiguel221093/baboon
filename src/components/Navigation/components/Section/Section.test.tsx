import { mount, shallow } from "enzyme";
import { NavigationContext } from "../../utils";
import { Item } from "../Item";
import Section from "./Section";

describe("Section", () => {
	const items = [
		{
			label: "Home",
			anchorProps: { href: "/" },
		},
		{
			label: "About",
			anchorProps: { href: "/about" },
		},
	];

	const title = "Main Navigation";

	const action = {
		id: "add",
		icon: "plus",
		onPress: jest.fn(),
		disabled: false,
		onMouseEnter: jest.fn(),
		onTouchStart: jest.fn(),
		loading: false,
		danger: false,
	};

	const contextValues = {
		collapsed: false,
	};

	it("renders correctly when there are items and a title", () => {
		const wrapper = shallow(
			<NavigationContext.Provider value={contextValues}>
				<Section items={items} title={title} />
			</NavigationContext.Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("renders correctly when there are items, a title, and an action", () => {
		const wrapper = shallow(
			<NavigationContext.Provider value={contextValues}>
				<Section items={items} title={title} action={action} />
			</NavigationContext.Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("renders correctly when there are no items and no title", () => {
		const wrapper = shallow(
			<NavigationContext.Provider value={contextValues}>
				<Section />
			</NavigationContext.Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("renders correctly when the navigation is collapsed", () => {
		const collapsedContextValues = {
			collapsed: true,
		};
		const wrapper = shallow(
			<NavigationContext.Provider value={collapsedContextValues}>
				<Section items={items} title={title} action={action} />
			</NavigationContext.Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("renders the correct number of items", () => {
		const wrapper = mount(
			<NavigationContext.Provider value={contextValues}>
				<Section items={items} />
			</NavigationContext.Provider>
		);
		expect(wrapper.find(Item)).toHaveLength(items.length);
	});
});
