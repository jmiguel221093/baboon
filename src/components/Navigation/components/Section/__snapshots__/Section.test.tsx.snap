// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section renders correctly when the navigation is collapsed 1`] = `ShallowWrapper {}`;

exports[`Section renders correctly when there are items and a title 1`] = `ShallowWrapper {}`;

exports[`Section renders correctly when there are items, a title, and an action 1`] = `ShallowWrapper {}`;

exports[`Section renders correctly when there are no items and no title 1`] = `ShallowWrapper {}`;
